inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/format/JsonDynamicSchemaFormat.java

```diff
@@ -0,0 +1,168 @@
+        if (physicalNode == null) {
+            for (String key : keys) {
+                values.add(extract(root, key));
+            }
+            return values;
+        }
+        for (String key : keys) {
+            String value = extract(physicalNode, key);
+            if (value == null) {
+                value = extract(root, key);
+            }
+            values.add(value);
+        }
+        return values;
+    }
+
+    /**
+     * Extract value by key from ${@link JsonNode}
+     *
+     * @param jsonNode The json node
+     * @param key The key that will be used to extract
+     * @return The value maps the key in the json node
+     */
+    @Override
+    public String extract(JsonNode jsonNode, String key) {
+        if (jsonNode == null || key == null) {
+            return null;
+        }
+        JsonNode value = jsonNode.get(key);
+        if (value != null) {
+            return value.asText();
+        }
+        int index = key.indexOf(".");
+        if (index > 0 && index + 1 < key.length()) {
+            return extract(jsonNode.get(key.substring(0, index)), key.substring(index + 1));
+        }
+        return null;
+    }
+
+    /**
+     * Deserialize from byte array and return a ${@link JsonNode}
+     *
+     * @param message The byte array of raw data
+     * @return The JsonNode
+     * @throws IOException The exceptions may throws when deserialize
+     */
+    @Override
+    public JsonNode deserialize(byte[] message) throws IOException {
+        return objectMapper.readTree(message);
+    }
+
+    /**
+     * Parse msg and replace the value by key from meta data and physical.
+     * See details {@link JsonDynamicSchemaFormat#parse(JsonNode, String)}
+     *
+     * @param message The source of data rows format by bytes
+     * @param pattern The pattern value
+     * @return The result of parsed
+     * @throws IOException The exception will throws
+     */
+    @Override
+    public String parse(byte[] message, String pattern) throws IOException {
+        return parse(deserialize(message), pattern);
+    }
+
+    /**
+     * Parse msg and replace the value by key from meta data and physical.
+     * Such as:
+     * 1. give a pattern "${a}{b}{c}" and the root Node contains the keys(a: '1', b: '2', c: '3')
+     * the result of pared will be '123'
+     * 2. give a pattern "${a}_{b}_{c}" and the root Node contains the keys(a: '1', b: '2', c: '3')
+     * the result of pared will be '1_2_3'
+     * 3. give a pattern "prefix_${a}_{b}_{c}_suffix" and the root Node contains the keys(a: '1', b: '2', c: '3')
+     * the result of pared will be 'prefix_1_2_3_suffix'
+     *
+     * @param rootNode The root node of json
+     * @param pattern The pattern value
+     * @return The result of parsed
+     * @throws IOException The exception will throws
+     */
+    @Override
+    public String parse(JsonNode rootNode, String pattern) throws IOException {
```

Comment:
Serach in `data` JsonNode  first then search in `root` JsonNode second？A little bit stange getting sink table from physicalData，what if there is a field named `database`  in physicalData？(╯﹏╰)b

Reply:
The essence of pattern extraction is to extract from data. It should be reasonable to distinguish between small physical fields and metadata fields. If there are both, physical fields are preferred.